src/Solana/Transaction.cpp

```diff
@@ -11,8 +11,80 @@
 #include "../BinaryCoding.h"
 #include "../PublicKey.h"
 
+#include <vector>
+
 using namespace TW;
 using namespace TW::Solana;
+using namespace std;
+
+uint8_t CompiledInstruction::findAccount(const Address& address) {
+    auto it = std::find(addresses.begin(), addresses.end(), address);
+    if (it == addresses.end()) {
+        throw new std::invalid_argument("address not found");
+    }
+    assert(it != addresses.end());
+    auto dist = std::distance(addresses.begin(), it);
+    assert(dist < 256);
+    return (uint8_t)dist;
+}
+
+void Message::addAccount(const AccountMeta& account) {
+    if (account.isSigner) {
+        if (std::find(signedAccounts.begin(), signedAccounts.end(), account.account) == signedAccounts.end()) {
+            signedAccounts.push_back(account.account);
+        }
+    } else if (!account.isReadOnly) {
+        if (std::find(signedAccounts.begin(), signedAccounts.end(), account.account) == signedAccounts.end() &&
+            std::find(unsignedAccounts.begin(), unsignedAccounts.end(), account.account) == unsignedAccounts.end()) {
+            unsignedAccounts.push_back(account.account);
+        }
+    } else {
+        if (std::find(signedAccounts.begin(), signedAccounts.end(), account.account) == signedAccounts.end() &&
+            std::find(unsignedAccounts.begin(), unsignedAccounts.end(), account.account) == unsignedAccounts.end() &&
+            std::find(readOnlyAccounts.begin(), readOnlyAccounts.end(), account.account) == readOnlyAccounts.end()) {
+            readOnlyAccounts.push_back(account.account);
+        }
+    }
+}
+
+void Message::compileAccounts() {
+    for (auto& instr: instructions) {
+        for (auto& address: instr.accounts) {
+            addAccount(address);
+        }
+    }
+    // add programIds (read-only)
+    for (auto& instr: instructions) {
```

Comment:
Are you really need two cycles?

Reply:
Yes, order is important (to match existing real transactions)